concrete-test/src/synthesizing.rs

```diff
@@ -0,0 +1,96 @@
+        $(
+            #[doc = "A trait to synthesize "]
+            #[doc = $kind]
+            #[doc = " from prototypical entities."]
+            pub trait $name<Prec: IntegerPrecision>: $entity < $($left=$right),* > where Maker: Prototyper<Prec>{
+                fn from_prototype(synth: &mut Maker, prototype: &<Maker as Prototyper<Prec>>::$prototype) -> Self;
+                fn into_prototype(synth: &mut Maker, entity: &Self) -> <Maker as Prototyper<Prec>>::$prototype;
+            }
+        )*
+    };
+}
+
+synthesizable_traits! {
+    {"cleartext", SynthesizableCleartext, CleartextEntity, CleartextPrototype},
+    {"cleartext vector", SynthesizableCleartextVector, CleartextVectorEntity, CleartextVectorPrototype},
+    {"plaintext", SynthesizablePlaintext, PlaintextEntity, PlaintextPrototype},
+    {"plaintext vector", SynthesizablePlaintextVector, PlaintextVectorEntity, PlaintextVectorPrototype}
+}
+
+synthesizable_traits! {
+    {"binary lwe secret key", SynthesizableBinaryLweSecretKey, LweSecretKeyEntity<KeyDistribution=BinaryKeyDistribution>, BinaryLweSecretKeyPrototype},
+    {"binary lwe ciphertext", SynthesizableBinaryLweCiphertext, LweCiphertextEntity<KeyDistribution=BinaryKeyDistribution>, BinaryLweCiphertextPrototype},
+    {"binary lwe ciphertext vector", SynthesizableBinaryLweCiphertextVector, LweCiphertextVectorEntity<KeyDistribution=BinaryKeyDistribution>, BinaryLweCiphertextVectorPrototype},
+    {"binary glwe secret key", SynthesizableBinaryGlweSecretKey, GlweSecretKeyEntity<KeyDistribution=BinaryKeyDistribution>, BinaryGlweSecretKeyPrototype},
+    {"binary glwe ciphertext", SynthesizableBinaryGlweCiphertext, GlweCiphertextEntity<KeyDistribution=BinaryKeyDistribution>, BinaryGlweCiphertextPrototype},
+    {"binary glwe ciphertext vector", SynthesizableBinaryGlweCiphertextVector, GlweCiphertextVectorEntity<KeyDistribution=BinaryKeyDistribution>, BinaryGlweCiphertextVectorPrototype},
+    {"binary to binary lwe keyswitch key", SynthesizableBinaryBinaryLweKeyswitchKey, LweKeyswitchKeyEntity<InputKeyDistribution=BinaryKeyDistribution, OutputKeyDistribution=BinaryKeyDistribution>, BinaryBinaryLweKeyswitchKeyPrototype},
+    {"binary to binary lwe bootstrap key", SynthesizableBinaryBinaryLweBootstrapKey, LweBootstrapKeyEntity<InputKeyDistribution=BinaryKeyDistribution, OutputKeyDistribution=BinaryKeyDistribution>, BinaryBinaryLweBootstrapKeyPrototype}
+}
+
+macro_rules! impl_synthesizable_for_core {
+    ($({$traitname:ident, $precision:ident, $entity:ident, $proto:ident}),*) => {
+        $(
+            impl $traitname<$precision> for $entity {
+                fn from_prototype(_synth: &mut Maker, prototype: &$proto) -> Self {
+                    prototype.0.to_owned()
+                }
+
+                fn into_prototype(_synth: &mut Maker, entity: &Self) -> $proto {
+                    $proto(entity.to_owned())
+                }
+            }
+        )*
+    };
+}
+
+impl_synthesizable_for_core! {
```

Comment:
Shouldn't this be gated by a `#[cfg(feature = "backend_core")]` ? since it uses types coming from the core engine.

Reply:
If the `core` backend is not present, there is no ways to make the tests work (any of them), so I don't think it would make sense to have a flag which when deactivated avoid the crate to build. In the end I guess the real question here is whether we would like to gate the `concrete-test::backends::core` module or not.

Reply:
I guess that for now, I will keep it that way, and if it ever comes to be a problem later on, we will change it. Thanks for the feedback !